File: 2/js/src/YoungestCommonAncestor.ts
class Vertex {
	data: String
	ancestor: Vertex | null

	constructor(data: string) {
		this.data = data;
		this.ancestor= null;
	}
}

function addVertex(vertices: Vertex[], vertexData: string, ancestorData?: string) {
	let vertex = getVertex(vertices, vertexData);
	if(vertex == null) {
		vertex = new Vertex(vertexData);
		vertices.push(vertex);
	}

	if(ancestorData != undefined) {
		let ancestor = getVertex(vertices, ancestorData);
		if(ancestor == null) {
			ancestor = new Vertex(ancestorData);
			vertices.push(ancestor);
		}

		vertex.ancestor = ancestor;
	}
}

function getVertex(vertices: Vertex[], data: string): Vertex | null {
	for(const vertex of vertices) {
		if(vertex.data == data) return vertex;
	}

	return null;
}

function getVertexDepth(vertex: Vertex): number {
	let depth = 0;

	let cur = vertex;
	while(cur.ancestor !=  null) {
		depth++;
		cur = cur.ancestor;
	}

	return depth;
}

function ascendToSameDepth(vertex: Vertex, diff: number): Vertex {
	while(diff > 0) {
		vertex = vertex.ancestor!;
		diff--;
	}

	return vertex;
}

function getYoungestCommonAncestor(vertices: Vertex[], descendantOne: Vertex, descendantTwo: Vertex) {
	const depthOne = getVertexDepth(descendantOne);
	const depthTwo = getVertexDepth(descendantTwo);

	const diff = Math.abs(depthOne - depthTwo);

	if(depthOne > depthTwo) {
		descendantOne = ascendToSameDepth(descendantOne, diff);
	} else {
		descendantTwo = ascendToSameDepth(descendantTwo, diff);
	}

	if(descendantOne == descendantTwo) {
		return descendantOne;
	}

	while(descendantOne.ancestor != descendantTwo.ancestor) {
		descendantOne = descendantOne.ancestor!;
		descendantTwo = descendantTwo.ancestor!;
	}

	return descendantOne.ancestor;
}

(function main() {
	const vertices: Vertex[] = [];

	addVertex(vertices, "B", "A");
	addVertex(vertices, "C", "A");
	addVertex(vertices, "D", "A");
	addVertex(vertices, "E", "A");
	addVertex(vertices, "F", "A");

	addVertex(vertices, "G", "B");
	addVertex(vertices, "H", "B");
	addVertex(vertices, "I", "B");

	addVertex(vertices, "J", "C");

	addVertex(vertices, "K", "D");
	addVertex(vertices, "L", "D");

	addVertex(vertices, "M", "F");
	addVertex(vertices, "N", "F");

	addVertex(vertices, "O", "H");
	addVertex(vertices, "P", "H");
	addVertex(vertices, "Q", "H");
	addVertex(vertices, "R", "H");

	addVertex(vertices, "T", "P");
	addVertex(vertices, "U", "P");

	addVertex(vertices, "V", "R");

	addVertex(vertices, "W", "V");
	addVertex(vertices, "X", "V");
	addVertex(vertices, "Y", "V");

	addVertex(vertices, "Z", "X");

	const descendantOne = getVertex(vertices, "Y");
	const descendantTwo = getVertex(vertices, "F");

	console.log(getYoungestCommonAncestor(vertices, descendantOne!, descendantTwo!));
})();

File: 2/java/YoungestCommonAncestor.java
import java.util.ArrayList;
import java.util.List;

class YoungestCommonAncestor {
	List<Node> nodes = new ArrayList<Node>();

	class Node {
		String data;
		Node ancestor = null;

		Node(String data) {
			this.data = data;
		}
	}
	
	public void setOldestAncestor(String data) {
		Node ancestor = new Node(data);
		this.nodes.add(ancestor);
	}
	
	public Node getNode(String data) {
		for(Node node : this.nodes) {
			if(node.data.equals(data)) {
				return node;
			}
		}
		return null;
	}

	public void addNode(String childData, String ancestorData) {
		Node ancestor = null;
		Node child = null;

		for(Node node : this.nodes) {
			if(node.data.equals(ancestorData)) {
				ancestor = node;
			}

			if(node.data.equals(childData)) {
				child = node;
			}
		}
		
		if(ancestor == null) {
			ancestor = new Node(ancestorData);
			this.nodes.add(ancestor);
		}

		if(child == null) {
			child = new Node(childData);
			this.nodes.add(child);
		}

		child.ancestor = ancestor;
	}	
	
	public int getLevel(Node topAncestor, Node descendant) {
		int level = 0;
		
		while(descendant != topAncestor) {
			descendant = descendant.ancestor;
			level++;
		}

		return level;
	}
	
	public Node getSameLevelNode(Node node, int lowerDepth, int upperDepth) {
		
		while(upperDepth > lowerDepth) {
			upperDepth--;
			node = node.ancestor;
		}	
		return node;
	}

	public Node getYoungestCommonAncestor(Node topAncestor, Node descendantOne, Node descendantTwo) {
		int depthOne = getLevel(topAncestor, descendantOne);
		int depthTwo = getLevel(topAncestor, descendantTwo);
		
		if(depthOne < depthTwo) {
			descendantTwo = getSameLevelNode(descendantTwo, depthOne, depthTwo); 
		} else if(depthOne > depthTwo) {
				descendantOne = getSameLevelNode(descendantOne, depthTwo, depthOne);
		}
			
		while(descendantOne.ancestor != descendantTwo.ancestor) {
			descendantOne = descendantOne.ancestor;
			descendantTwo = descendantTwo.ancestor;
		}		
		
		if(descendantOne == descendantTwo) {
			return descendantOne;
		}

		return descendantOne.ancestor;
	}


	public static void main(String args[]) {
		YoungestCommonAncestor y = new YoungestCommonAncestor();
		y.setOldestAncestor("A");

		y.addNode("B", "A");
		y.addNode("C", "A");
		y.addNode("D", "A");
		y.addNode("E", "A");
		y.addNode("F", "A");

		y.addNode("G", "B");
		y.addNode("H", "B");
		y.addNode("I", "B");

		y.addNode("J", "C");
		
		y.addNode("K", "D");
		y.addNode("L", "D");

		y.addNode("M", "F");
		y.addNode("N", "F");

		y.addNode("O", "H");
		y.addNode("P", "H");
		y.addNode("Q", "H");
		y.addNode("R", "H");

		y.addNode("T", "P");
		y.addNode("U", "P");

		y.addNode("V", "R");

		y.addNode("W", "V");
		y.addNode("X", "V");
		y.addNode("Y", "V");

		y.addNode("Z", "X");
		
		Node A = y.getNode("A"); 
		Node descendantOne = y.getNode("V");
		Node descendantTwo = y.getNode("Z");

		System.out.println(y.getYoungestCommonAncestor(A, descendantOne, descendantTwo).data);
	}

}
